File: backend/app/api/v1/jobs.py
"""
backend/app/api/v1/jobs.py
─────────────────────────────────────────────────────────────────────────────
Canonical REST API routes for unified Job tracking, status polling, and cancellation.
─────────────────────────────────────────────────────────────────────────────
"""

import logging
from typing import Optional, List
from fastapi import APIRouter, HTTPException, Query, Depends

from api.dependencies.auth import get_current_user
from services.jobs import job_manager, JobRecord, JobStatusResponse, JobListResponse

logger = logging.getLogger("sonikoma.api.jobs")

jobs_router = APIRouter()

@jobs_router.get(
    "/{job_id}",
    response_model=JobStatusResponse,
    summary="Get job status, progress, stage, execution, and result",
    description="Returns the full execution state of a specific job including provider, model, attempt, stage, progress, project_id, and chapter_id."
)
async def get_job_status_endpoint(job_id: str, current_user: dict = Depends(get_current_user)):
    """Canonical job status retrieval endpoint."""
    job = job_manager.get_job(job_id)
    if not job:
        raise HTTPException(status_code=404, detail=f"Job '{job_id}' not found.")
    if job.user_id != current_user["user_id"]:
        raise HTTPException(status_code=403, detail="Not authorized to access this job.")
    return job.to_status_response()

@jobs_router.post(
    "/{job_id}/cancel",
    response_model=JobStatusResponse,
    summary="Cancel a running or queued job",
    description="Cancels an active background execution task."
)
async def cancel_job_endpoint(job_id: str, current_user: dict = Depends(get_current_user)):
    """Cancels a running or queued job."""
    job = job_manager.get_job(job_id)
    if not job:
        raise HTTPException(status_code=404, detail=f"Job '{job_id}' not found.")
    if job.user_id != current_user["user_id"]:
        raise HTTPException(status_code=403, detail="Not authorized to cancel this job.")
    job = job_manager.cancel_job(job_id)
    return job.to_status_response()

@jobs_router.get(
    "/",
    response_model=JobListResponse,
    summary="List jobs with filtering and pagination",
    description="Retrieves a paginated list of background execution jobs for the authenticated user, filterable by project_id, chapter_id, status, and job_type."
)
async def list_jobs_endpoint(
    project_id: Optional[str] = Query(None, description="Filter by parent Project/Series ID"),
    chapter_id: Optional[str] = Query(None, description="Filter by Chapter/Episode ID"),
    status: Optional[str] = Query(None, description="Filter by status: QUEUED, RUNNING, COMPLETED, FAILED, CANCELLED"),
    job_type: Optional[str] = Query(None, description="Filter by Job Type (e.g. SCRAPE_CHAPTER, PANEL_SPLIT, GENERATE_STORYBOARD, RENDER_VIDEO)"),
    limit: int = Query(50, ge=1, le=200, description="Max number of jobs to return"),
    offset: int = Query(0, ge=0, description="Pagination offset"),
    current_user: dict = Depends(get_current_user)
):
    """Lists jobs, optionally filtered by project_id, chapter_id, status, and type."""
    jobs = job_manager.list_jobs(
        user_id=current_user["user_id"],
        project_id=project_id,
        chapter_id=chapter_id,
        status=status,
        job_type=job_type,
        limit=limit,
        offset=offset
    )
    return JobListResponse(
        success=True,
        total=len(jobs),
        jobs=[j.to_status_response() for j in jobs],
    )

File: frontend/src/api/endpoints/jobs.ts
/**
 * frontend/src/api/endpoints/jobs.ts
 * ─────────────────────────────────────────────────────────────────────────────
 * Canonical Job Status Polling & Management API Client
 * ─────────────────────────────────────────────────────────────────────────────
 */

export interface JobExecutionInfo {
  provider?: string;
  model?: string;
  attempt?: number;
}

export interface JobErrorInfo {
  code: string;
  message: string;
  stage?: string;
  provider?: string;
  model?: string;
}

export interface JobStatusResponse<T = any> {
  job_id: string;
  job_type: string;
  capability?: string;
  status: "queued" | "running" | "completed" | "failed" | "cancelled";
  progress: number;
  stage: string;
  project_id?: string | null;
  chapter_id?: string | null;
  execution?: JobExecutionInfo | null;
  result?: T | null;
  error?: JobErrorInfo | null;
}

/**
 * Retrieves the current execution state of a job from the canonical endpoint.
 */
export async function getJobStatus<T = any>(
  jobId: string
): Promise<JobStatusResponse<T>> {
  const token =
    localStorage.getItem("sonikoma_token") ||
    sessionStorage.getItem("sonikoma_token") ||
    "";

  const res = await fetch(`/api/jobs/${jobId}`, {
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
  });

  if (!res.ok) {
    throw new Error(`Failed to fetch job status (HTTP ${res.status})`);
  }

  return res.json();
}

/**
 * Cancels a running or queued job.
 */
export async function cancelJob<T = any>(
  jobId: string
): Promise<JobStatusResponse<T>> {
  const token =
    localStorage.getItem("sonikoma_token") ||
    sessionStorage.getItem("sonikoma_token") ||
    "";

  const res = await fetch(`/api/jobs/${jobId}/cancel`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
  });

  if (!res.ok) {
    throw new Error(`Failed to cancel job (HTTP ${res.status})`);
  }

  return res.json();
}

/**
 * Polls a job until completion or failure with callbacks.
 */
export async function pollJob<T = any>(
  jobId: string,
  options: {
    intervalMs?: number;
    timeoutMs?: number;
    onProgress?: (job: JobStatusResponse<T>) => void;
  } = {}
): Promise<JobStatusResponse<T>> {
  const intervalMs = options.intervalMs || 1000;
  const timeoutMs = options.timeoutMs || 300000; // 5 min timeout
  const startTime = Date.now();

  while (Date.now() - startTime < timeoutMs) {
    const job = await getJobStatus<T>(jobId);
    options.onProgress?.(job);

    if (job.status === "completed") {
      return job;
    }

    if (job.status === "failed") {
      const err = new Error(
        job.error?.message || `Job ${jobId} failed at stage ${job.stage}`
      ) as any;
      err.job = job;
      throw err;
    }

    if (job.status === "cancelled") {
      const err = new Error(`Job ${jobId} was cancelled`) as any;
      err.job = job;
      throw err;
    }

    await new Promise((resolve) => setTimeout(resolve, intervalMs));
  }

  throw new Error(`Job ${jobId} timed out after ${timeoutMs}ms`);
}

export interface JobListResponse<T = any> {
  success: boolean;
  total: number;
  jobs: JobStatusResponse<T>[];
}

export interface ListJobsOptions {
  projectId?: string;
  chapterId?: string;
  status?: string;
  jobType?: string;
  limit?: number;
  offset?: number;
}

/**
 * Lists background execution jobs with optional filtering and pagination.
 */
export async function listJobs<T = any>(
  options: ListJobsOptions = {}
): Promise<JobListResponse<T>> {
  const token =
    localStorage.getItem("sonikoma_token") ||
    sessionStorage.getItem("sonikoma_token") ||
    "";

  const params = new URLSearchParams();
  if (options.projectId) params.append("project_id", options.projectId);
  if (options.chapterId) params.append("chapter_id", options.chapterId);
  if (options.status) params.append("status", options.status);
  if (options.jobType) params.append("job_type", options.jobType);
  if (options.limit) params.append("limit", String(options.limit));
  if (options.offset) params.append("offset", String(options.offset));

  const qs = params.toString();
  const url = `/api/jobs/${qs ? `?${qs}` : ""}`;
  const res = await fetch(url, {
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
  });

  if (!res.ok) {
    throw new Error(`Failed to list jobs (HTTP ${res.status})`);
  }

  return res.json();
}
